### importing libraries

In [5]:
#importing libraries
import pandas as pd
import numpy as np
import os

### Importing data

In [6]:
df = pd.read_csv(r'C:\Users\amany\A4-Instacart Basket Analysis\02 Data\original data\orders.csv', index_col=False)
path = r'C:\Users\amany\A4-Instacart Basket Analysis'
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original data', 'products.csv'), index_col = False)
df_ords = pd.read_csv(r'C:\Users\amany\A4-Instacart Basket Analysis\02 Data\original data\orders.csv')
df_deps = pd.read_csv(r'C:\Users\amany\A4-Instacart Basket Analysis\02 Data\original data\departments.csv')


In [3]:
#create a data frame
df_test = pd.DataFrame()

In [4]:
#creat a mixed type column
df_test['mix'] = ['a', 'b', 1, True]

In [5]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [6]:
#chack for mixed types
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [7]:
#find missing values
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [8]:
# view these 16 missed values
df_nan = df_prods[df_prods['product_name'].isnull() == True]
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [9]:
#find missing values for orders
df_ords.isnull().sum()

order_id                       0
user_id                        0
eval_set                       0
order_number                   0
order_dow                      0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

### addressing missed values

In [10]:
#use the mean 
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

In [11]:
#replace the missing values
#df['product_name'].fillna(mean value, inplace=True)

In [12]:
#check the current number of rows in your

df_prods.shape

(49693, 5)

In [13]:
#use the mean 
df_ords.describe()

order_id       user_id  order_number     order_dow  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01  2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01  2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00  1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01  3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01  5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02  6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

In [14]:
# remove the missing values entirely from products df
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [15]:
#check the new number of rows in your

df_prods_clean.shape

(49677, 5)

In [16]:
#Another way you can drop all missing values is via the following command:
#df_prods.dropna(subset = [‘product_name’], inplace = True)



In [17]:
#replace the missing values with mean in orders df
df['days_since_prior_order'].fillna(1.111484e+01
, inplace=True)

In [18]:
#find missing values for orders
df_ords.isnull().sum()

order_id                       0
user_id                        0
eval_set                       0
order_number                   0
order_dow                      0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

### duplicated data

In [19]:
#The following command will look for full duplicates within your dataframe:
df_dups = df_prods_clean[df_prods_clean.duplicated()]


In [20]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [21]:
#check the current number of rows in your
df_prods_clean.shape

(49677, 5)

In [22]:
#remove duplicated data
df_prods.drop_duplicates()

product_id                                       product_name  \
0               1                         Chocolate Sandwich Cookies   
1               2                                   All-Seasons Salt   
2               3               Robust Golden Unsweetened Oolong Tea   
3               4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4               5                          Green Chile Anytime Sauce   
...           ...                                                ...   
49688       49684          Vodka, Triple Distilled, Twist of Vanilla   
49689       49685                 En Croute Roast Hazelnut Cranberry   
49690       49686                                   Artisan Baguette   
49691       49687         Smartblend Healthy Metabolism Dry Cat Food   
49692       49688                             Fresh Foaming Cleanser   

       aisle_id  department_id  prices  
0            61             19     5.8  
1           104             13     9.3  
2            94              7     4.5  
3            38              1    10.5  
4             5             13     4.3  
...         ...            ...     ...  
49688       124              5     5.3  
49689        42              1     3.1  
49690       112              3     7.8  
49691        41              8     4.7  
49692        73             11    13.5  

[49688 rows x 5 columns]

In [23]:
#check the new number of rows in df after removing duplicated data
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()


In [24]:
df_prods_clean_no_dups.shape

(49672, 5)

## working on the orders df

In [25]:
df_ords.describe()

order_id       user_id  order_number     order_dow  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01  2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01  2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00  1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01  3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01  5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02  6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

In [26]:
#create a data frame
df_ords_test = pd.DataFrame()

In [27]:
#creat a mixed type column
df_ords_test['mix'] = ['a', 'b', 1, True]

In [28]:
df_ords_test.head()

mix
0     a
1     b
2     1
3  True

In [29]:
#chack for mixed types
for col in df_ords_test.columns.tolist():
  weird = (df_ords_test[[col]].applymap(type) != df_ords_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords_test[weird]) > 0:
    print (col)

mix


In [30]:
#find missing values
df_ords.isnull().sum()

order_id                       0
user_id                        0
eval_set                       0
order_number                   0
order_dow                      0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

In [50]:
print(df_ords)

         order_id  user_id eval_set  order_number  order_dow  \
0         2539329        1    prior             1          2   
1         2398795        1    prior             2          3   
2          473747        1    prior             3          3   
3         2254736        1    prior             4          4   
4          431534        1    prior             5          4   
...           ...      ...      ...           ...        ...   
3421078   2266710   206209    prior            10          5   
3421079   1854736   206209    prior            11          4   
3421080    626363   206209    prior            12          1   
3421081   2977660   206209    prior            13          1   
3421082    272231   206209    train            14          6   

         order_hour_of_day  days_since_prior_order  
0                        8                     0.0  
1                        7                    15.0  
2                       12                    21.0  
3                  

In [32]:
#Find out the the missing values based on user_id count summarized 
df_ords[df_ords['days_since_prior_order'].isnull()].groupby('user_id').count().describe()

order_id  eval_set  order_number  order_dow  order_hour_of_day  \
count  206209.0  206209.0      206209.0   206209.0           206209.0   
mean        1.0       1.0           1.0        1.0                1.0   
std         0.0       0.0           0.0        0.0                0.0   
min         1.0       1.0           1.0        1.0                1.0   
25%         1.0       1.0           1.0        1.0                1.0   
50%         1.0       1.0           1.0        1.0                1.0   
75%         1.0       1.0           1.0        1.0                1.0   
max         1.0       1.0           1.0        1.0                1.0   

       days_since_prior_order  
count                206209.0  
mean                      0.0  
std                       0.0  
min                       0.0  
25%                       0.0  
50%                       0.0  
75%                       0.0  
max                       0.0

#### there are about 206209 missing values in days_since_prior_order column, while the the df_ords dataframe has a total of 3421083 rows and 7 columns, the total number of missed values is 206209.

#Address the missing values using mean function , mean =0

In [33]:
df_ords['days_since_prior_order'].fillna(0, inplace=True)


In [34]:
#make sure that missed values replaced
df_ords.isnull().sum()

order_id                  0
user_id                   0
eval_set                  0
order_number              0
order_dow                 0
order_hour_of_day         0
days_since_prior_order    0
dtype: int64

now the orders data set dont contain any missed/incorrect values, i used the mean method bcz its numeric value.


### working on duplicated data

In [35]:
# look for full duplicates within your dataframe:
#df_dups = df_ords_clean[df_ords_clean.duplicated()]
df_dups = df_ords[df_ords.duplicated()]


In [36]:
df_dups

Empty DataFrame
Columns: [order_id, user_id, eval_set, order_number, order_dow, order_hour_of_day, days_since_prior_order]
Index: []

there is no dublicated data in orders data frame.

In [37]:
df_ords.columns


Index(['order_id', 'user_id', 'eval_set', 'order_number', 'order_dow',
       'order_hour_of_day', 'days_since_prior_order'],
      dtype='object')

In [38]:
df_ords

order_id  user_id eval_set  order_number  order_dow  \
0         2539329        1    prior             1          2   
1         2398795        1    prior             2          3   
2          473747        1    prior             3          3   
3         2254736        1    prior             4          4   
4          431534        1    prior             5          4   
...           ...      ...      ...           ...        ...   
3421078   2266710   206209    prior            10          5   
3421079   1854736   206209    prior            11          4   
3421080    626363   206209    prior            12          1   
3421081   2977660   206209    prior            13          1   
3421082    272231   206209    train            14          6   

         order_hour_of_day  days_since_prior_order  
0                        8                     0.0  
1                        7                    15.0  
2                       12                    21.0  
3                        7                    29.0  
4                       15                    28.0  
...                    ...                     ...  
3421078                 18                    29.0  
3421079                 10                    30.0  
3421080                 12                    18.0  
3421081                 12                     7.0  
3421082                 14                    30.0  

[3421083 rows x 7 columns]

In [39]:
df_prods_clean_no_dups.describe()

product_id      aisle_id  department_id        prices
count  49672.000000  49672.000000   49672.000000  49672.000000
mean   24850.349775     67.762442      11.728942      9.993282
std    14340.705287     38.315784       5.850779    453.615536
min        1.000000      1.000000       1.000000      1.000000
25%    12432.750000     35.000000       7.000000      4.100000
50%    24850.500000     69.000000      13.000000      7.100000
75%    37268.250000    100.000000      17.000000     11.100000
max    49688.000000    134.000000      21.000000  99999.000000

#### Export your final, cleaned df_prods and df_ords data as “.csv” files in your “Prepared Data” folder and give them appropriate, succinct names.

In [48]:
# Export the cleaned df_prods dataframe
df_prods.to_csv(os.path.join(path, '02 data', 'prepared data', 'products_cleaned.csv'), index=False)

In [49]:
# Export the cleaned df_ords dataframe
df_ords.to_csv(os.path.join(path, '02 Data', 'Prepared data', 'orders_cleaned.csv'), index=False)